src/main/java/uk/gov/pay/ledger/transaction/service/AccountIdListSupplierManager.java

```diff
@@ -0,0 +1,53 @@
+package uk.gov.pay.ledger.transaction.service;
+
+import uk.gov.pay.ledger.exception.ValidationException;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.lang.String.format;
+import static org.apache.commons.lang3.ObjectUtils.isEmpty;
+
+public class AccountIdListSupplierManager<T> {
+
+    private static final String GATEWAY_ACCOUNT_ID = "account_id";
+
+    private final boolean overrideAccountRestriction;
+    private final List<String> gatewayAccountIds;
+
+    private Supplier<T> privilegedSupplier;
+    private Function<List<String>, T> supplier;
+
+    public AccountIdListSupplierManager(Boolean overrideAccountRestriction, List<String> gatewayAccountIds) {
+        this.overrideAccountRestriction = Optional.ofNullable(overrideAccountRestriction).orElse(false);
+        this.gatewayAccountIds = gatewayAccountIds;
+    }
+
+    public static <U> AccountIdListSupplierManager<U> of(Boolean overrideAccountRestriction, List<String> gatewayAccountId) {
+        return new AccountIdListSupplierManager<>(overrideAccountRestriction, gatewayAccountId);
+    }
+
+    public AccountIdListSupplierManager<T> withPrivilegedSupplier(Supplier<T> supplier) {
+        this.privilegedSupplier = supplier;
+        return this;
+    }
+
+    public AccountIdListSupplierManager<T> withSupplier(Function<List<String>, T> supplier) {
+        this.supplier = supplier;
+        return this;
+    }
+
+    public T validateAndGet() {
+        if(!overrideAccountRestriction && isEmpty(gatewayAccountIds)) {
```

Comment:
Did you really mean to use `org.apache.commons.lang3.ObjectUtils.isEmpty` here? If you want the `null` check, I would just do…

```java
if (!overrideAccountRestriction && (gatewayAccountIds == null || gatewayAccountIds.isEmpty())) {
```

Reply:
does `gatewayAccountIds == null || gatewayAccountIds.isEmpty()` not throw an NPE?

Reply:
If `gatewayAccountIds == null` it won’t evaluate the other side because it already knows that the whole logical or expression will return `true`.